### PR TITLE
Core execution refactor, EIP 684, edge case fixes, starting nonce for morden

### PR DESCRIPTION
**Pull request by austinabell**
- Refactors CALL code executions to not all rely on the same function and perform additional logic
- Implements EIP 684
- Fixes all edge case tests from #13 
- Fixes starting nonce logic for Morden

This does not completely solve the issue from https://github.com/etclabscore/go-ethereum/issues/12 but now the discrepancy is just a gas usage differential and the post state root hashes are matching up with the tx. These changes pass all state tests and I will remove the invalid state tests existing in the repository before that expect the incorrect behaviour and investigate the test failures to try to finish off these tests before opening PR for review.

### CODE REVIEW

**Comment by soc1c** [comment]
how much is required to finalize this?

**Comment by austinabell** [comment]
> how much is required to finalize this?

Just debugging the gas usage discrepancy, I can give context on how this is very convoluted in the codebase but it shouldn't be a difficult fix. I was waiting to merge in this PR once I solved the consensus bug, and most (I can't say all yet) of the classic failing tests currently are outdated by EIP684 impl

**Comment by austinabell** [comment]
Removed failing tests since they are both cases where it as assumed a contract collision will overwrite a contract. PR is ready for review now. Solves all state test discrepancies and the post state root hash was matching on the Morden tx now and was just giving a gas usage error for the tx (and tx receipt because of this) so that issue can be assumed to be seperate from this PR, unless the issue is found before this gets merged in. Closes #13 by implementing final difference, EIP684

**Comment by soc1c** [comment]
I'm stuck 

```
~/.opt/getc-core austin/executionrefactor
❯ bin/geth --testnet --rpc
2019-08-04 16:16:08 Geth Classic version: v6.0.6-23-ge7b4d8b
2019-08-04 16:16:08 Blockchain: Morden Testnet
2019-08-04 16:16:08 Chain database: /home/user/.ethereum-classic/morden/chaindata
2019-08-04 16:16:08 Blockchain upgrades configured: 7
2019-08-04 16:16:08   494000 Homestead
2019-08-04 16:16:08  1783000 GasReprice
2019-08-04 16:16:08  1885000 The DAO Hard Fork
2019-08-04 16:16:08  1915000 Diehard
2019-08-04 16:16:08  2000000 Gotham
2019-08-04 16:16:08  2300000 Defuse Difficulty Bomb
2019-08-04 16:16:08  4729274 Atlantis
2019-08-04 16:16:08 Using 21 configured bootnodes
2019-08-04 16:16:08 Use Sputnik EVM: false
2019-08-04 16:16:08 No etherbase set and no accounts found as default
2019-08-04 16:16:08 Allotted 1024MB cache and 1024 file handles to /home/user/.ethereum-classic/morden/chaindata
2019-08-04 16:16:10 Allotted 16MB cache and 16 file handles to /home/user/.ethereum-classic/morden/dapp
2019-08-04 16:16:10 Protocol Versions: [63 62], Network Id: 2, Chain Id: 62
2019-08-04 16:16:10 Genesis block: 0x0cd786a2425d16f152c658316c423e6ce1181e15c3295826d7c9904cba9ce303 (morden testnet)
2019-08-04 16:16:10 Local head header:     #4729796 [0xc6c4a3…] TD=624501362029068
2019-08-04 16:16:10 Local head full block: #4729796 [0xc6c4a3…] TD=624501362029068
2019-08-04 16:16:10 Local head fast block: #4729796 [0xc6c4a3…] TD=624501362029068
2019-08-04 16:16:10 Starting server...
2019-08-04 16:16:10 Mapping network port udp:30303 -> 30303 (ethereum discovery) using NAT-PMP(192.168.0.1)
2019-08-04 16:16:10 UDP listening. Client enode: enode://d529883f7d66a96a98e0dab15371d8b58d84d81796c7ae28333faf49635e24348e2aecc26c33cb8dcd390129bf1a2af46e81e0eabd09d691b302df08429078b6@192.168.1.100:30303
2019-08-04 16:16:10 HTTP endpoint: http://localhost:8545
2019-08-04 16:16:10 Debug log config: verbosity=5 log-dir=/home/user/.ethereum-classic/morden/log vmodule=*
2019-08-04 16:16:10 Display log config: display=3 status=sync=1m0s
2019-08-04 16:16:10 Machine log config: mlog=off
2019-08-04 16:16:10 IPC endpoint opened: /home/user/.ethereum-classic/morden/geth.ipc
2019-08-04 16:16:10 Mapping network port tcp:30303 -> 30303 (ethereum p2p) using NAT-PMP(192.168.0.1)
2019-08-04 16:17:10 Discover                                                                 0/25 peers
2019-08-04 16:18:10 Discover                                                                 0/25 peers
2019-08-04 16:19:10 Sync      #4729796 of  #4766182 c6c4a38b     0/   0/ 0 blk/txs/mgas sec  1/25 peers
2019-08-04 16:20:10 Sync      #4729796 of  #4766182 c6c4a38b     0/   0/ 0 blk/txs/mgas sec  3/25 peers
2019-08-04 16:21:10 Sync      #4729796 of  #4730535 c6c4a38b     0/   0/ 0 blk/txs/mgas sec  2/25 peers
2019-08-04 16:22:10 Sync      #4729796 of  #4766183 c6c4a38b     0/   0/ 0 blk/txs/mgas sec  3/25 peers
```

**Review by soc1c — CHANGES_REQUESTED** [review]
I'm sorry, this breaks Morden already on block `323`.

```
2019-08-04 20:36:46 Discover                                                                 1/25 peers
2019-08-04 20:37:46 Sync          #323 of  #4767414 6de28537     0/   0/ 0 blk/txs/mgas sec  2/25 peers
2019-08-04 20:38:46 Sync          #323 of  #4767415 6de28537     0/   0/ 0 blk/txs/mgas sec  2/25 peers
2019-08-04 20:39:46 Sync          #323 of  #4767416 6de28537     0/   0/ 0 blk/txs/mgas sec  3/25 peers
2019-08-04 20:40:46 Sync          #323 of  #4767419 6de28537     0/   0/ 0 blk/txs/mgas sec  3/25 peers
2019-08-04 20:41:46 Sync          #323 of  #4767423 6de28537     0/   0/ 0 blk/txs/mgas sec  3/25 peers
2019-08-04 20:42:46 Sync          #323 of  #4767426 6de28537     0/   0/ 0 blk/txs/mgas sec  3/25 peers
2019-08-04 20:43:46 Sync          #323 of  #4767427 6de28537     0/   0/ 0 blk/txs/mgas sec  3/25 peers
2019-08-04 20:44:46 Discover                                                                 1/25 peers
2019-08-04 20:45:46 Discover                                                                 1/25 peers
2019-08-04 20:46:46 Discover                                                                 1/25 peers
2019-08-04 20:47:46 Discover                                                                 1/25 peers
2019-08-04 20:48:46 Discover                                                                 1/25 peers
2019-08-04 20:49:46 Discover                                                                 1/25 peers
2019-08-04 20:50:46 Discover                                                                 1/25 peers
2019-08-04 20:51:46 Sync          #323 of  #4767470 6de28537     0/   0/ 0 blk/txs/mgas sec  2/25 peers
2019-08-04 20:52:46 Sync          #323 of  #4767471 6de28537     0/   0/ 0 blk/txs/mgas sec  2/25 peers
2019-08-04 20:53:46 Discover                                                                 1/25 peers
2019-08-04 20:54:46 Sync          #323 of  #4767482 6de28537     0/   0/ 0 blk/txs/mgas sec  2/25 peers
2019-08-04 20:55:46 Sync          #323 of  #4767488 6de28537     0/   0/ 0 blk/txs/mgas sec  2/25 peers
2019-08-04 20:56:46 Sync          #323 of  #4767490 6de28537     0/   0/ 0 blk/txs/mgas sec  2/25 peers
2019-08-04 20:57:46 Sync          #323 of  #4767495 6de28537     0/   0/ 0 blk/txs/mgas sec  2/25 peers
2019-08-04 20:58:46 Sync          #323 of  #4767501 6de28537     0/   0/ 0 blk/txs/mgas sec  2/25 peers
2019-08-04 20:59:46 Sync          #323 of  #4767504 6de28537     0/   0/ 0 blk/txs/mgas sec  4/25 peers
2019-08-04 21:00:46 Sync          #323 of  #4767510 6de28537     0/   0/ 0 blk/txs/mgas sec  3/25 peers
2019-08-04 21:01:46 Sync          #323 of  #4767516 6de28537     0/   0/ 0 blk/txs/mgas sec  3/25 peers
2019-08-04 21:02:46 Sync          #323 of  #4767521 6de28537     0/   0/ 0 blk/txs/mgas sec  3/25 peers
2019-08-04 21:03:46 Discover                                                                 1/25 peers
2019-08-04 21:04:46 Discover                                                                 1/25 peers
2019-08-04 21:05:46 Discover                                                                 1/25 peers
2019-08-04 21:06:46 Discover                                                                 1/25 peers
2019-08-04 21:07:46 Discover                                                                 1/25 peers
2019-08-04 21:08:46 Discover                                                                 1/25 peers
2019-08-04 21:09:46 Discover                                                                 1/25 peers
2019-08-04 21:10:46 Sync          #323 of  #4767565 6de28537     0/   0/ 0 blk/txs/mgas sec  2/25 peers
2019-08-04 21:11:46 Sync          #323 of  #4767568 6de28537     0/   0/ 0 blk/txs/mgas sec  2/25 peers
2019-08-04 21:12:46 Sync          #323 of  #4767576 6de28537     0/   0/ 0 blk/txs/mgas sec  2/25 peers
2019-08-04 21:13:46 Sync          #323 of  #4767578 6de28537     0/   0/ 0 blk/txs/mgas sec  3/25 peers
2019-08-04 21:14:46 Sync          #323 of  #4767583 6de28537     0/   0/ 0 blk/txs/mgas sec  3/25 peers
2019-08-04 21:15:46 Sync          #323 of  #4767587 6de28537     0/   0/ 0 blk/txs/mgas sec  3/25 peers
2019-08-04 21:16:46 Sync          #323 of  #4767590 6de28537     0/   0/ 0 blk/txs/mgas sec  3/25 peers
```

**Comment by austinabell** [comment]
I will look into it early tomorrow but I don't see how this is possible, I have synced up many times with this branch and had just done so before opening PR for review.

**Review by steviezhang — CHANGES_REQUESTED** [review]
can confirm gets stuck at block 323.

opcodes from the contract creation tx in the next block (pulled from parity) here
https://pastebin.com/Xpa4pMtu

**Comment by steviezhang** [comment]
@austinabell @soc1c we're stuck on block 323 because block 324 contains the first contract creation transaction and our EIP684 explicitly checks for a nonce of 0 but accounts on morden start with a nonce of 2^20

**Comment by soc1c** [comment]
Good find, I'm resyncing now.

**Comment by austinabell** [comment]
resolves #22, if anyone else can double check

**Review by soc1c — APPROVED** [review]
works finally.

**Comment by GregTheGreek** [comment]
Syncing... code checks out

**Comment by GregTheGreek** [comment]
@r8d8 @tzdybal Hey please review :)

**Comment by soc1c** [comment]
Screenshot at 2019-08-07 12-53-54

**Review by steviezhang — APPROVED** [review]

File: tests/state_test_util.go [hunk 1]
```diff
@@ -200,7 +200,7 @@ func runStateTest(ruleSet RuleSet, test VmTest) error {
 		if balance, ok := new(big.Int).SetString(account.Balance, 0); !ok {
 			panic("malformed test account balance")
 		} else if balance.Cmp(obj.Balance()) != 0 {
-			return fmt.Errorf("(%x) balance failed. Expected: %v have: %v\n", obj.Address().Bytes()[:4], account.Balance, obj.Balance())
+			return fmt.Errorf("(%x) balance failed. Expected: %v have: %v\n", obj.Address().Bytes()[:4], account.Balance, "0x"+obj.Balance().Text(16))
```
Comment (tzdybal) [inline tests/state_test_util.go:203]: You can easily add "0x" inside error format string, not as a concatenation in argument.

**Review by GregTheGreek — APPROVED** [review]

**Comment by steviezhang** [comment]
Maybe rename it to include the starting nonce details/have it in the description

File: core/execution.go [hunk 2]
```diff
@@ -17,53 +17,217 @@
...
+	}
 	return ret, err
 }
 
 // CallCode executes the given address' code as the given contract address
 func CallCode(env vm.Environment, caller vm.ContractRef, addr common.Address, input []byte, gas, gasPrice, value *big.Int) (ret []byte, err error) {
-	callerAddr := caller.Address()
-	ret, _, err = exec(env, caller, &callerAddr, &addr, env.Db().GetCodeHash(addr), input, env.Db().GetCode(addr), gas, gasPrice, value, false)
+	evm := env.Vm()
+	// Depth check execution. Fail if we're trying to execute above the limit.
```
Comment (tzdybal) [inline core/execution.go:106]: This few lines look like a code duplicate, function could be extracted.

File: core/execution.go [hunk 3]
```diff
@@ -17,53 +17,217 @@
...
+	}
 	return ret, err
 }
 
 // StaticCall executes within the given contract and throws exception if state is attempted to be changed
 func StaticCall(env vm.Environment, caller vm.ContractRef, addr common.Address, input []byte, gas, gasPrice *big.Int) (ret []byte, err error) {
-	ret, _, err = exec(env, caller, &addr, &addr, env.Db().GetCodeHash(addr), input, env.Db().GetCode(addr), gas, gasPrice, new(big.Int), true)
+	evm := env.Vm()
+	// Depth check execution. Fail if we're trying to execute above the limit.
+	if env.Depth() > callCreateDepthMax {
```
Comment (tzdybal) [inline core/execution.go:186]: Another code duplicate

**Review by tzdybal — APPROVED** [review]
LGTM. I found only some small "stylistic" issues.
